core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -101,6 +101,41 @@ public static Snapshot oldestAncestor(Table table) {
     return ancestorsOf(start, lookup);
   }
 
+  /**
+   * Traverses the history of the table's current snapshot
+   * and finds the oldest ancestor snapshot after or equal to the timestamp in milliseconds.
+   * @return null if there is no current snapshot in the table,
+   * else the oldest ancestor snapshot after or equal to the timestamp in milliseconds.
+   */
+  public static Snapshot oldestAncestorAfter(Table table, Long timestampMillis) {
+    Snapshot current = table.currentSnapshot();
+    long timestamp = timestampMillis == null ? -1L : timestampMillis;
+    if (current == null || current.timestampMillis() < timestamp) {
+      return null;
+    }
+
+    for (Snapshot snapshot : currentAncestors(table)) {
+      if (snapshot.timestampMillis() < timestamp) {
+        break;
+      }
+
+      current = snapshot;
+    }
+
+    return current;
```

Comment:
I don't think this logic is correct. In the version that I wrote, the method will throw an `IllegalStateException` if it cannot determine whether the ancestor was actually the first snapshot after the given timestamp. It is fine if the behavior you want is to ignore that case in practice, but the utility methods in `SnapshotUtil` should be more careful. What you should do instead is call `oldestAncestor` if this fails to find one and you want to default to the oldest ancestor.

Reply:
I check the logic from `snapshotAfter`, and compare the logic with the current function `oldestAncestorAfter`, I think the current implementation is correct. In current implementation, current will point the snapshot which has the timestamp >= targetTimestamp, the loop will break when current snapshot points to the first available one that is greater than the targetTimestamp. It is not like the `snapshotAfter` implementation, it is potential that there is no `current.parentId() != snapshotId`, we throw IllegalStateException.
The implementation of  `oldestAncestorAfter` will try its best effort to find the 1st snapshot greater than targetTimestamp. If it can't find, the result will be the oldestAncestor while the loop function exits. I don't see any difference as call the `oldestAncestor` if it fails to find one(result should be the same with current implementation). Please let me know if I miss something here.
But for safety purpose, I will add  `oldestAncestor` calls last when it fails to find one, thanks for the suggestion.